src/violation.rs

```diff
@@ -42,16 +51,38 @@ impl<'a> ViolationBuilder<'a> {
         self.fix = Some(Some(Rc::new(callback)));
         self
     }
+
+    pub fn message(&mut self, message: impl Into<String>) -> &mut Self {
+        let message = message.into();
+        self.message_or_message_id = Some(MessageOrMessageId::Message(message));
+        self
+    }
+
+    pub fn message_id(&mut self, message_id: impl Into<String>) -> &mut Self {
+        let message_id = message_id.into();
+        self.message_or_message_id = Some(MessageOrMessageId::MessageId(message_id));
+        self
+    }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
+pub enum MessageOrMessageId {
+    Message(String),
+    MessageId(String),
+}
+
+pub type ViolationData = HashMap<String, String>;
+
+#[derive(Clone, Debug)]
 pub struct ViolationWithContext {
-    pub message: String,
+    pub message_or_message_id: MessageOrMessageId,
     pub range: tree_sitter::Range,
     pub path: PathBuf,
     pub rule: RuleMeta,
     pub plugin_index: Option<PluginIndex>,
     pub was_fix: bool,
+    pub kind: &'static str,
```

Comment:
Don't know where I land wrt `type`/`type_` vs `kind` for referring to "the type of tree-sitter AST nodes"